code/game/objects/items/devices/aicard.dm

```diff
@@ -37,71 +37,78 @@
+			if(law in AI.laws.ion_laws) // If we're an ion law, give it an ion index code
+				laws.Add(ionnum() + ". " + law.law)
+			else
+				laws.Add(num2text(law.get_index()) + ". " + law.law)
 		data["laws"] = laws
-		data["has_laws"] = laws.len
+		data["has_laws"] = length(AI.laws.all_laws())
+
+	else
+		data["has_ai"] = FALSE // If this isn't passed to tgui, it won't show there isn't a AI in the card.
 
 	return data
 
 
-/obj/item/aicard/Topic(href, href_list, nowindow, state)
+/obj/item/aicard/tgui_act(action, params)
 	if(..())
-		return 1
+		return
 
 	var/mob/living/silicon/ai/AI = locate() in src
 	if(!istype(AI))
-		return 1
+		return
 
 	var/user = usr
-
-	if(href_list["wipe"])
-		var/confirm = alert("Are you sure you want to wipe this card's memory? This cannot be undone once started.", "Confirm Wipe", "Yes", "No")
-		if(confirm == "Yes" && (CanUseTopic(user, state) == STATUS_INTERACTIVE))
+	switch(action)
+		if("wipe")
 			msg_admin_attack("[key_name_admin(user)] wiped [key_name_admin(AI)] with \the [src].", ATKLOG_FEW)
 			add_attack_logs(user, AI, "Wiped with [src].")
-			flush = 1
-			AI.suiciding = 1
-			to_chat(AI, "Your core files are being wiped!")
-			while(AI && AI.stat != DEAD)
-				AI.adjustOxyLoss(2)
-				sleep(10)
-			flush = 0
-
-	if(href_list["radio"])
-		AI.aiRadio.disabledAi = text2num(href_list["radio"])
-		to_chat(AI, "<span class='warning'>Your Subspace Transceiver has been [AI.aiRadio.disabledAi ? "disabled" : "enabled"]!</span>")
-		to_chat(user, "<span class='notice'>You [AI.aiRadio.disabledAi ? "disable" : "enable"] the AI's Subspace Transceiver.</span>")
-
-	if(href_list["wireless"])
-		AI.control_disabled = text2num(href_list["wireless"])
-		to_chat(AI, "<span class='warning'>Your wireless interface has been [AI.control_disabled ? "disabled" : "enabled"]!</span>")
-		to_chat(user, "<span class='notice'>You [AI.control_disabled ? "disable" : "enable"] the AI's wireless interface.</span>")
-		update_icon()
-
-	return 1
+			INVOKE_ASYNC(src, .proc/wipe_ai)
+
+		if("radio")
+			AI.aiRadio.disabledAi = !AI.aiRadio.disabledAi
+			to_chat(AI, "<span class='warning'>Your Subspace Transceiver has been [AI.aiRadio.disabledAi ? "disabled" : "enabled"]!</span>")
+			to_chat(user, "<span class='notice'>You [AI.aiRadio.disabledAi ? "disable" : "enable"] the AI's Subspace Transceiver.</span>")
+
+		if("wireless")
+			AI.control_disabled = !AI.control_disabled
+			to_chat(AI, "<span class='warning'>Your wireless interface has been [AI.control_disabled ? "disabled" : "enabled"]!</span>")
+			to_chat(user, "<span class='notice'>You [AI.control_disabled ? "disable" : "enable"] the AI's wireless interface.</span>")
+			update_icon()
+
+	return TRUE
+
+/obj/item/aicard/proc/wipe_ai()
+	var/mob/living/silicon/ai/AI = locate() in src
+	flush = TRUE
+	AI.suiciding = TRUE
+	to_chat(AI, "Your core files are being wiped!")
+	while(AI && AI.stat != DEAD)
+		AI.adjustOxyLoss(2)
+		sleep(10)
+	flush = FALSE
```

Comment:
May be worth adding a check here so you cant spam this button
```
if(flush)
    to_chat(user, "<span class='warning'>Youre already wiping this AI</span>")
```
etc

Reply:
handled by disabling the wipe button when wiping
`disabled={data.flushing || data.integrity === 0}`

Reply:
Just marking the button as disabled isnt adequate here. **NEVER** trust a clients input since it is modifiable at runtime, and **ALWAYS** have hardcoded safety checks server side.

Reply:
True. One moment.